src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpContentHeaders.cs

```diff
@@ -67,33 +67,61 @@ public ICollection<string> ContentLanguage
         {
             get
             {
-                // 'Content-Length' can only hold one value. So either we get 'null' back or a boxed long value.
-                object? storedValue = GetParsedValues(KnownHeaders.ContentLength.Descriptor);
-
-                // Only try to calculate the length if the user didn't set the value explicitly using the setter.
-                if (!_contentLengthSet && (storedValue == null))
+                // This could use GetParsedValues; however, for responses we typically access ContentLength
+                // only once or twice: once as part of the handler, and potentially once as part of the
+                // consumer, e.g. HttpClient.GetStringAsync.  In such cases, we're better off just parsing
+                // the raw response content-length twice and not storing anything back, as today storing it
+                // back incurs multiple allocations, on top of which if the response headers are enumerated,
+                // we then need to format the boxed length back into a string.  Given typical usage patterns,
+                // it's better to just parse each time and pay the few additional nanoseconds to re-parse.
```

Comment:
I might add a comment here to the following effect:

If parsing multiple times ever becomes a perf issue, then the best way to handle it would be to cache the Content-Length value here, on HttpContentHeaders, as a `long`. This would require some additional coordination with the underlying HttpHeaders header store stuff, in case someone then called something like Add or Remove on Content-Length at the HttpHeaders level. But it's certainly doable; it's probably as a simple as a callback to HttpContentHeaders that says "Content-Length has been modified in the store, please discard any cached value".